File: modules/iris-look-bake/android/src/main/java/expo/modules/irislookbake/AnimeStylize.kt
package expo.modules.irislookbake

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import androidx.exifinterface.media.ExifInterface
import java.io.File
import java.io.FileOutputStream
import java.nio.FloatBuffer
import java.util.UUID
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

object AnimeStylize {
  private val STYLE_TO_MODEL = mapOf(
    "animegan-v3-shinkai" to "AnimeGANv3_Shinkai_37.onnx",
    "animegan-v3-hayao" to "AnimeGANv3_Hayao_36.onnx",
  )
  private const val MAX_SIDE = 1024
  private const val ALIGN = 8
  private const val MIN_SIDE = 256

  private val sessions = mutableMapOf<String, OrtSession>()
  /** Serializes session create + Run — ORT sessions are not safe for concurrent Run. */
  private val runLock = Any()
  private val env: OrtEnvironment = OrtEnvironment.getEnvironment()

  fun stylizePhoto(
    context: Context?,
    inputPath: String,
    style: String,
    strength: Double,
  ): Map<String, Any> {
    val modelAsset = STYLE_TO_MODEL[style]
      ?: throw IllegalArgumentException("Unsupported style: $style")
    val ctx = context ?: throw IllegalStateException("No Android context")
    val strengthClamped = strength.coerceIn(0.0, 1.0)
    val path = stripFileScheme(inputPath)

    synchronized(runLock) {
      val source = decodeOrientedBitmap(path)
        ?: throw IllegalStateException("Could not decode photo for anime stylize")

      try {
        if (strengthClamped <= 0.01) {
          return passthrough(path)
        }

        val origW = source.width
        val origH = source.height
        val (inferW, inferH) = alignedSize(origW, origH, MAX_SIDE)
        val resized = Bitmap.createScaledBitmap(source, inferW, inferH, true)
        var stylizedBitmap: Bitmap? = null
        var fullRes: Bitmap? = null
        var sharpened: Bitmap? = null
        var blended: Bitmap? = null

        try {
          val input = bitmapToNHWC(resized)
          val ortSession = sharedSessionLocked(ctx, modelAsset)
          val inputName = ortSession.inputNames.firstOrNull()
            ?: throw IllegalStateException("AnimeGANv3 model has no input names")
          val inputShape = longArrayOf(1, inferH.toLong(), inferW.toLong(), 3)
          val tensor = OnnxTensor.createTensor(env, FloatBuffer.wrap(input), inputShape)

          tensor.use {
            ortSession.run(mapOf(inputName to it)).use { results ->
              val outTensor = results[0] as? OnnxTensor
                ?: throw IllegalStateException("Anime inference returned unexpected tensor type")
              val count = inferW * inferH * 3
              val remaining = outTensor.floatBuffer.remaining()
              if (remaining < count) {
                throw IllegalStateException(
                  "Anime inference output too short ($remaining < $count)",
                )
              }
              val out = FloatArray(count)
              outTensor.floatBuffer.duplicate().get(out)
              stylizedBitmap = nhwcToBitmap(out, inferW, inferH)
            }
          }

          val stylized = stylizedBitmap
            ?: throw IllegalStateException("Anime inference produced no bitmap")
          fullRes = Bitmap.createScaledBitmap(stylized, origW, origH, true)
          sharpened = sharpen(fullRes!!, 0.65f)
          blended = if (strengthClamped >= 0.999) {
            sharpened
          } else {
            blend(source, sharpened!!, strengthClamped.toFloat())
          }

          val outFile = writeJpeg(ctx, blended!!)
          return mapOf(
            "path" to outFile.absolutePath,
            "uri" to outFile.toURI().toString(),
          )
        } finally {
          recycleDistinct(resized, source)
          recycleDistinct(stylizedBitmap, fullRes, sharpened, blended, source)
          recycleDistinct(fullRes, sharpened, blended, source)
          recycleDistinct(sharpened, blended, source)
          recycleDistinct(blended, source)
        }
      } finally {
        if (!source.isRecycled) source.recycle()
      }
    }
  }

  private fun passthrough(path: String): Map<String, Any> {
    val file = File(path)
    return mapOf(
      "path" to file.absolutePath,
      "uri" to file.toURI().toString(),
    )
  }

  private fun stripFileScheme(path: String): String =
    if (path.startsWith("file://")) path.removePrefix("file://") else path

  /** Decode JPEG/PNG and apply EXIF orientation (parity with iOS UIImage). */
  private fun decodeOrientedBitmap(path: String): Bitmap? {
    val decoded = BitmapFactory.decodeFile(path) ?: return null
    return try {
      val exif = ExifInterface(path)
      val orientation = exif.getAttributeInt(
        ExifInterface.TAG_ORIENTATION,
        ExifInterface.ORIENTATION_NORMAL,
      )
      if (orientation == ExifInterface.ORIENTATION_NORMAL ||
        orientation == ExifInterface.ORIENTATION_UNDEFINED
      ) {
        return decoded
      }
      val matrix = Matrix()
      when (orientation) {
        ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.preScale(-1f, 1f)
        ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
        ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.preScale(1f, -1f)
        ExifInterface.ORIENTATION_TRANSPOSE -> {
          matrix.postRotate(90f)
          matrix.preScale(-1f, 1f)
        }
        ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
        ExifInterface.ORIENTATION_TRANSVERSE -> {
          matrix.postRotate(270f)
          matrix.preScale(-1f, 1f)
        }
        ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
        else -> return decoded
      }
      val oriented = Bitmap.createBitmap(decoded, 0, 0, decoded.width, decoded.height, matrix, true)
      if (oriented !== decoded) decoded.recycle()
      oriented
    } catch (_: Exception) {
      decoded
    }
  }

  /** Scale to maxSide, then align each dim to ×8 with min 256 (AnimeGANv3). */
  private fun alignedSize(width: Int, height: Int, maxSide: Int): Pair<Int, Int> {
    val scale = min(1.0, maxSide.toDouble() / max(width, height).toDouble())
    var tw = (width * scale).roundToInt()
    var th = (height * scale).roundToInt()
    tw = max(MIN_SIDE, ((tw + ALIGN - 1) / ALIGN) * ALIGN)
    th = max(MIN_SIDE, ((th + ALIGN - 1) / ALIGN) * ALIGN)
    return tw to th
  }

  /** Caller must hold [runLock]. */
  private fun sharedSessionLocked(context: Context, modelAsset: String): OrtSession {
    sessions[modelAsset]?.let { return it }
    val bytes = context.assets.open(modelAsset).use { it.readBytes() }
    val options = OrtSession.SessionOptions()
    // CPU only — NNAPI (like CoreML EP) is unsafe for these dynamic H×W models.
    val created = env.createSession(bytes, options)
    sessions[modelAsset] = created
    return created
  }

  private fun bitmapToNHWC(bitmap: Bitmap): FloatArray {
    val w = bitmap.width
    val h = bitmap.height
    val pixels = IntArray(w * h)
    bitmap.getPixels(pixels, 0, w, 0, 0, w, h)
    val out = FloatArray(w * h * 3)
    for (i in pixels.indices) {
      val c = pixels[i]
      val base = i * 3
      out[base] = ((c shr 16) and 0xff) / 127.5f - 1f
      out[base + 1] = ((c shr 8) and 0xff) / 127.5f - 1f
      out[base + 2] = (c and 0xff) / 127.5f - 1f
    }
    return out
  }

  private fun nhwcToBitmap(floats: FloatArray, width: Int, height: Int): Bitmap {
    val pixels = IntArray(width * height)
    for (i in pixels.indices) {
      val base = i * 3
      val r = u8((floats[base] + 1f) * 127.5f)
      val g = u8((floats[base + 1] + 1f) * 127.5f)
      val b = u8((floats[base + 2] + 1f) * 127.5f)
      pixels[i] = (0xff shl 24) or (r shl 16) or (g shl 8) or b
    }
    return Bitmap.createBitmap(pixels, width, height, Bitmap.Config.ARGB_8888)
  }

  private fun u8(v: Float): Int = max(0, min(255, v.roundToInt()))

  /** Mild unsharp via 3×3 kernel (amount ~0…1). */
  private fun sharpen(src: Bitmap, amount: Float): Bitmap {
    val a = amount.coerceIn(0f, 1f)
    if (a < 0.01f) return src
    val w = src.width
    val h = src.height
    val pixels = IntArray(w * h)
    src.getPixels(pixels, 0, w, 0, 0, w, h)
    val out = IntArray(w * h)
    // center = 1 + 4a, neighbors = -a
    val c = 1f + 4f * a
    val n = -a
    for (y in 0 until h) {
      for (x in 0 until w) {
        fun sample(sx: Int, sy: Int): Int {
          val xx = sx.coerceIn(0, w - 1)
          val yy = sy.coerceIn(0, h - 1)
          return pixels[yy * w + xx]
        }
        val p0 = sample(x, y)
        val pL = sample(x - 1, y)
        val pR = sample(x + 1, y)
        val pU = sample(x, y - 1)
        val pD = sample(x, y + 1)
        fun ch(shift: Int): Int {
          val v =
            ((p0 shr shift) and 0xff) * c +
              ((pL shr shift) and 0xff) * n +
              ((pR shr shift) and 0xff) * n +
              ((pU shr shift) and 0xff) * n +
              ((pD shr shift) and 0xff) * n
          return max(0, min(255, v.roundToInt()))
        }
        out[y * w + x] = (0xff shl 24) or (ch(16) shl 16) or (ch(8) shl 8) or ch(0)
      }
    }
    return Bitmap.createBitmap(out, w, h, Bitmap.Config.ARGB_8888)
  }

  private fun blend(original: Bitmap, stylized: Bitmap, strength: Float): Bitmap {
    val w = original.width
    val h = original.height
    val out = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
    val canvas = Canvas(out)
    val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    canvas.drawBitmap(original, 0f, 0f, paint)
    paint.alpha = (strength.coerceIn(0f, 1f) * 255f).roundToInt()
    canvas.drawBitmap(stylized, 0f, 0f, paint)
    return out
  }

  private fun writeJpeg(context: Context, bitmap: Bitmap): File {
    val dir = File(context.filesDir, "looks")
    if (!dir.exists()) dir.mkdirs()
    val out = File(dir, "iris-anime-${UUID.randomUUID()}.jpg")
    FileOutputStream(out).use { stream ->
      if (!bitmap.compress(Bitmap.CompressFormat.JPEG, 95, stream)) {
        throw IllegalStateException("Could not encode anime JPEG")
      }
    }
    return out
  }

  private fun recycleDistinct(candidate: Bitmap?, vararg keep: Bitmap?) {
    if (candidate == null || candidate.isRecycled) return
    if (keep.any { it != null && it === candidate }) return
    candidate.recycle()
  }
}

File: modules/iris-look-bake/ios/AnimeStylize.swift
import CoreGraphics
import CoreImage
import Foundation
import UIKit
import onnxruntime_objc

enum AnimeStylizeError: LocalizedError {
  case invalidStyle
  case modelMissing
  case decodeFailed
  case encodeFailed
  case inferenceFailed(String)

  var errorDescription: String? {
    switch self {
    case .invalidStyle: return "Unsupported anime style"
    case .modelMissing: return "AnimeGANv3 model missing from app bundle"
    case .decodeFailed: return "Could not decode photo for anime stylize"
    case .encodeFailed: return "Could not encode anime JPEG"
    case .inferenceFailed(let message): return "Anime inference failed: \(message)"
    }
  }
}

/// On-device AnimeGANv3 (Shinkai / Hayao) via ONNX Runtime (CPU).
/// CoreML EP is intentionally disabled: these models have dynamic H×W and crash under CoreML EP.
enum AnimeStylize {
  private static let styleToModel: [String: String] = [
    "animegan-v3-shinkai": "AnimeGANv3_Shinkai_37",
    "animegan-v3-hayao": "AnimeGANv3_Hayao_36",
  ]
  private static let maxSide = 1024
  private static let align = 8
  private static let minSide = 256
  /// Serializes session create + Run — ORT sessions are not safe for concurrent Run.
  private static let runLock = NSLock()
  private static var cachedSessions: [String: ORTSession] = [:]
  private static var cachedEnv: ORTEnv?
  private static let ciContext = CIContext(options: [.useSoftwareRenderer: false])

  static func stylizePhoto(inputPath: String, style: String, strength: Double) throws -> [String: Any] {
    guard let modelFileName = styleToModel[style] else { throw AnimeStylizeError.invalidStyle }

    let strength = max(0, min(1, strength))
    let inputURL = URL(fileURLWithPath: stripFileScheme(inputPath))
    guard let source = loadOrientedCGImage(at: inputURL.path) else {
      throw AnimeStylizeError.decodeFailed
    }

    if strength <= 0.01 {
      return passthrough(url: inputURL)
    }

    runLock.lock()
    defer { runLock.unlock() }

    let origW = source.width
    let origH = source.height
    let (inferW, inferH) = alignedSize(width: origW, height: origH, maxSide: maxSide)

    guard let resized = resize(source, width: inferW, height: inferH) else {
      throw AnimeStylizeError.decodeFailed
    }

    let inputTensor = try floatTensorNHWC(from: resized)
    let session = try sharedSessionLocked(modelFileName: modelFileName)

    let output: ORTValue
    do {
      let inputNameList = try session.inputNames()
      guard let inputName = inputNameList.first else {
        throw AnimeStylizeError.inferenceFailed("No input names")
      }
      let outputNameList = try session.outputNames()
      guard let firstName = outputNameList.first else {
        throw AnimeStylizeError.inferenceFailed("No output names")
      }
      let outputs = try session.run(
        withInputs: [inputName: inputTensor],
        outputNames: Set([firstName]),
        runOptions: nil
      )
      guard let value = outputs[firstName] else {
        throw AnimeStylizeError.inferenceFailed("No output tensor")
      }
      output = value
    } catch let err as AnimeStylizeError {
      throw err
    } catch {
      throw AnimeStylizeError.inferenceFailed(String(describing: error))
    }

    guard let stylized = image(fromNHWC: output, width: inferW, height: inferH) else {
      throw AnimeStylizeError.inferenceFailed("Output tensor shape mismatch or decode failed")
    }

    guard let fullRes = resize(stylized, width: origW, height: origH) else {
      throw AnimeStylizeError.decodeFailed
    }
    // Mild clarity after upscale — recovers edge snap lost in infer→full-res.
    let sharpened = sharpen(fullRes, intensity: 0.65) ?? fullRes

    let blended: CGImage
    if strength >= 0.999 {
      blended = sharpened
    } else if let mix = blend(original: source, stylized: sharpened, strength: CGFloat(strength)) {
      blended = mix
    } else {
      blended = sharpened
    }

    let outURL = try writeJPEG(blended, quality: 0.95)
    return [
      "path": outURL.path,
      "uri": outURL.absoluteString,
    ]
  }

  private static func passthrough(url: URL) -> [String: Any] {
    ["path": url.path, "uri": url.absoluteString]
  }

  private static func stripFileScheme(_ path: String) -> String {
    if path.hasPrefix("file://") {
      return URL(string: path)?.path ?? path.replacingOccurrences(of: "file://", with: "")
    }
    return path
  }

  /// Decode with EXIF orientation applied (VisionCamera masters are often rotated).
  private static func loadOrientedCGImage(at path: String) -> CGImage? {
    guard let ui = UIImage(contentsOfFile: path) else { return nil }
    if let cg = ui.cgImage, ui.imageOrientation == .up {
      return cg
    }
    let format = UIGraphicsImageRendererFormat.default()
    format.scale = 1
    format.opaque = true
    let size = ui.size
    let renderer = UIGraphicsImageRenderer(size: size, format: format)
    let flattened = renderer.image { _ in
      ui.draw(in: CGRect(origin: .zero, size: size))
    }
    return flattened.cgImage
  }

  /// Scale to maxSide, then align each dim to ×8 with min 256 (AnimeGANv3).
  private static func alignedSize(width: Int, height: Int, maxSide: Int) -> (Int, Int) {
    let scale = min(1.0, Double(maxSide) / Double(max(width, height)))
    var tw = Int((Double(width) * scale).rounded())
    var th = Int((Double(height) * scale).rounded())
    tw = max(minSide, ((tw + align - 1) / align) * align)
    th = max(minSide, ((th + align - 1) / align) * align)
    return (tw, th)
  }

  private static func modelPath(modelFileName: String) throws -> String {
    let host = Bundle(for: IrisLookBakeModule.self)
    let candidates: [Bundle] = [
      Bundle.main.url(forResource: "IrisLookBake", withExtension: "bundle").flatMap(Bundle.init(url:)),
      host.url(forResource: "IrisLookBake", withExtension: "bundle").flatMap(Bundle.init(url:)),
      host,
      Bundle.main,
    ].compactMap { $0 }

    for bundle in candidates {
      if let path = bundle.path(forResource: modelFileName, ofType: "onnx") {
        return path
      }
    }
    throw AnimeStylizeError.modelMissing
  }

  /// Caller must hold [runLock].
  private static func sharedSessionLocked(modelFileName: String) throws -> ORTSession {
    if let cached = cachedSessions[modelFileName] { return cached }

    let env: ORTEnv
    if let cachedEnv {
      env = cachedEnv
    } else {
      env = try ORTEnv(loggingLevel: .warning)
      cachedEnv = env
    }
    let options = try ORTSessionOptions()
    // CPU only — CoreML EP SEGV on this dynamic-shape AnimeGANv3 ONNX.
    try options.setGraphOptimizationLevel(.basic)
    try options.setIntraOpNumThreads(2)
    let session = try ORTSession(
      env: env,
      modelPath: try modelPath(modelFileName: modelFileName),
      sessionOptions: options
    )
    cachedSessions[modelFileName] = session
    return session
  }

  private static func resize(_ image: CGImage, width: Int, height: Int) -> CGImage? {
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let ctx = CGContext(
      data: nil,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: width * 4,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else { return nil }
    ctx.interpolationQuality = .high
    ctx.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
    return ctx.makeImage()
  }

  private static func sharpen(_ image: CGImage, intensity: Float) -> CGImage? {
    let input = CIImage(cgImage: image)
    guard let filter = CIFilter(name: "CISharpenLuminance") else { return nil }
    filter.setValue(input, forKey: kCIInputImageKey)
    filter.setValue(intensity, forKey: kCIInputSharpnessKey)
    guard let output = filter.outputImage else { return nil }
    let rect = CGRect(x: 0, y: 0, width: image.width, height: image.height)
    return ciContext.createCGImage(output, from: rect)
  }

  private static func floatTensorNHWC(from image: CGImage) throws -> ORTValue {
    let width = image.width
    let height = image.height
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    var pixels = [UInt8](repeating: 0, count: width * height * 4)
    guard let ctx = CGContext(
      data: &pixels,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: width * 4,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else {
      throw AnimeStylizeError.decodeFailed
    }
    ctx.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))

    var floats = [Float](repeating: 0, count: width * height * 3)
    for i in 0..<(width * height) {
      let o = i * 4
      let base = i * 3
      // Undo premultiplied alpha roughly (alpha is 255 from our contexts).
      floats[base] = Float(pixels[o]) / 127.5 - 1.0
      floats[base + 1] = Float(pixels[o + 1]) / 127.5 - 1.0
      floats[base + 2] = Float(pixels[o + 2]) / 127.5 - 1.0
    }

    // Copy into owned NSMutableData — never pass &Array storage directly to ORT.
    let byteCount = floats.count * MemoryLayout<Float>.size
    let data = floats.withUnsafeBufferPointer { buf -> NSMutableData in
      NSMutableData(bytes: buf.baseAddress, length: byteCount)
    }
    let shape: [NSNumber] = [1, NSNumber(value: height), NSNumber(value: width), 3]
    return try ORTValue(
      tensorData: data,
      elementType: .float,
      shape: shape
    )
  }

  private static func image(fromNHWC value: ORTValue, width: Int, height: Int) -> CGImage? {
    guard let tensorData = try? value.tensorData() else { return nil }
    let count = width * height * 3
    let expectedBytes = count * MemoryLayout<Float>.size
    // Require exact element count — short buffers corrupt pixels; longer may be padded OK.
    guard tensorData.length >= expectedBytes else { return nil }

    let bytes = tensorData as Data
    let floats: [Float] = bytes.withUnsafeBytes { raw in
      let buffer = raw.bindMemory(to: Float.self)
      guard buffer.count >= count else { return [] }
      return Array(buffer.prefix(count))
    }
    guard floats.count == count else { return nil }

    var pixels = [UInt8](repeating: 255, count: width * height * 4)
    for i in 0..<(width * height) {
      let base = i * 3
      let o = i * 4
      pixels[o] = u8((floats[base] + 1.0) * 127.5)
      pixels[o + 1] = u8((floats[base + 1] + 1.0) * 127.5)
      pixels[o + 2] = u8((floats[base + 2] + 1.0) * 127.5)
      pixels[o + 3] = 255
    }

    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let ctx = CGContext(
      data: &pixels,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: width * 4,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else { return nil }
    return ctx.makeImage()
  }

  private static func u8(_ v: Float) -> UInt8 {
    UInt8(max(0, min(255, v.rounded())))
  }

  private static func blend(original: CGImage, stylized: CGImage, strength: CGFloat) -> CGImage? {
    let width = original.width
    let height = original.height
    guard stylized.width == width, stylized.height == height else { return stylized }
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    var a = [UInt8](repeating: 0, count: width * height * 4)
    var b = [UInt8](repeating: 0, count: width * height * 4)
    guard
      let ctxA = CGContext(
        data: &a, width: width, height: height, bitsPerComponent: 8,
        bytesPerRow: width * 4, space: colorSpace,
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
      ),
      let ctxB = CGContext(
        data: &b, width: width, height: height, bitsPerComponent: 8,
        bytesPerRow: width * 4, space: colorSpace,
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
      )
    else { return nil }
    ctxA.draw(original, in: CGRect(x: 0, y: 0, width: width, height: height))
    ctxB.draw(stylized, in: CGRect(x: 0, y: 0, width: width, height: height))

    let t = Float(strength)
    let u = 1 - t
    for i in 0..<(width * height * 4) {
      if i % 4 == 3 {
        a[i] = 255
        continue
      }
      a[i] = UInt8(max(0, min(255, (Float(a[i]) * u + Float(b[i]) * t).rounded())))
    }
    return ctxA.makeImage()
  }

  private static func writeJPEG(_ image: CGImage, quality: CGFloat) throws -> URL {
    let ui = UIImage(cgImage: image)
    guard let data = ui.jpegData(compressionQuality: quality) else {
      throw AnimeStylizeError.encodeFailed
    }
    let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
      .appendingPathComponent("looks", isDirectory: true)
    try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    let out = dir.appendingPathComponent("iris-anime-\(UUID().uuidString).jpg")
    try data.write(to: out, options: .atomic)
    return out
  }
}
